core/unix/os.c

```diff
@@ -4454,17 +4454,29 @@ fd_priv_dup(file_t curfd)
          * so how do we tell if the flag is supported?  try calling once at init?
          */
         newfd = fcntl_syscall(curfd, F_DUPFD, min_dr_fd);
-        if (newfd < 0) {
+        int try_fd = min_dr_fd / 2;
+        static const int MIN_FD = 10; /* Avoid tcsh, etc. issues: see comment below. */
+        while (newfd < 0 && try_fd >= MIN_FD) {
             /* We probably ran out of fds, esp if debug build and there are
              * lots of threads.  Should we track how many we've given out to
              * avoid a failed syscall every time after?
              */
-            SYSLOG_INTERNAL_WARNING_ONCE("ran out of stolen fd space");
+            SYSLOG_INTERNAL_WARNING_ONCE("ran out of stolen fd space with error %d",
+                                         newfd);
             /* Try again but this time in the app space, somewhere high up
              * to avoid issues like tcsh assuming it can own fds 3-5 for
              * piping std{in,out,err} (xref the old -open_tcsh_fds option).
              */
-            newfd = fcntl_syscall(curfd, F_DUPFD, min_dr_fd / 2);
+            newfd = fcntl_syscall(curfd, F_DUPFD, try_fd);
```

Comment:
> with ENOMEM returned by the first two DUPFD attempts in fd_priv_dup()

The man page for fcntl doesn't mention ENOMEM at all though: https://man7.org/linux/man-pages/man2/fcntl.2.html.

Reply:
The code here printed out 12 for the return value is all I know. Certainly man pages are not alway 100% accurate for every kernel.  Not sure what you're asking for here: yes it was definitely 12, which is ENOMEM.  It is a slightly custom kernel but it would not be surprising if even mainline kernels had behavior not in the man page regarding errors.

Reply:
Going to resolve.  The PR description saying ENOMEM does say "internal environments" if this is unique to our internal kernel.  It doesn't hurt other kernels to have some retries before failure even if others wouldn't fail in this way where a retry would succeed.